File: desktop/src/lib/api/arme.ts
import { apiClient } from "@/lib/api-client";
import type {
  ApiResponse,
  Arme,
  ArmeMunitionsConsommation,
  TypeArme,
} from "@/types";

// ========================
// TypeArme API (weapon type catalog)
// ========================

export interface TypeArmePayload {
  nom: string;
  description?: string | null;
  /** Shared ammunition stock for all weapons of this type. */
  munitions_stock?: number;
}

export async function getTypeArmeList(
  search?: string,
): Promise<TypeArme[]> {
  const params = new URLSearchParams();
  if (search) params.set("search", search);
  const query = params.toString();
  const { data } = await apiClient.get<ApiResponse<TypeArme[]>>(
    `/types-armes${query ? `?${query}` : ""}`,
  );
  return data.data;
}

export async function getTypeArmeById(id: number): Promise<TypeArme> {
  const { data } = await apiClient.get<ApiResponse<TypeArme>>(
    `/types-armes/${id}`,
  );
  return data.data;
}

export async function createTypeArme(
  payload: TypeArmePayload,
): Promise<TypeArme> {
  const { data } = await apiClient.post<ApiResponse<TypeArme>>(
    "/types-armes",
    payload,
  );
  return data.data;
}

export async function updateTypeArme(
  id: number,
  payload: Partial<TypeArmePayload>,
): Promise<TypeArme> {
  const { data } = await apiClient.put<ApiResponse<TypeArme>>(
    `/types-armes/${id}`,
    payload,
  );
  return data.data;
}

export async function deleteTypeArme(id: number): Promise<void> {
  await apiClient.delete(`/types-armes/${id}`);
}

// ========================
// Arme API (individual weapon instances + ammunition stock)
// ========================

export interface ArmePayload {
  type_arme_id: number;
  matricule: string;
  munitions_stock?: number;
}

export async function getArmeList(
  filters?: Record<string, string>,
): Promise<Arme[]> {
  const params = new URLSearchParams(filters);
  const query = params.toString();
  const { data } = await apiClient.get<ApiResponse<Arme[]>>(
    `/armes${query ? `?${query}` : ""}`,
  );
  return data.data;
}

export async function getArmeById(id: number): Promise<Arme> {
  const { data } = await apiClient.get<ApiResponse<Arme>>(
    `/armes/${id}`,
  );
  return data.data;
}

export async function createArme(payload: ArmePayload): Promise<Arme> {
  const { data } = await apiClient.post<ApiResponse<Arme>>(
    "/armes",
    payload,
  );
  return data.data;
}

export async function updateArme(
  id: number,
  payload: Partial<ArmePayload>,
): Promise<Arme> {
  const { data } = await apiClient.put<ApiResponse<Arme>>(
    `/armes/${id}`,
    payload,
  );
  return data.data;
}

export async function deleteArme(id: number): Promise<void> {
  await apiClient.delete(`/armes/${id}`);
}

// ========================
// Consommation API (ammunition consumption + history)
// ========================

export interface ConsommationPayload {
  agent_id?: number | null;
  quantite: number;
  armement_id?: number | null;
}

/** Records an ammunition consumption for a specific weapon.
 * The backend atomically decreases the weapon's stock and inserts
 * a consumption history row inside a single transaction. Returns
 * the updated weapon with its new stock value. */
export async function recordConsommation(
  armeId: number,
  payload: ConsommationPayload,
): Promise<Arme> {
  const { data } = await apiClient.post<ApiResponse<Arme>>(
    `/armes/${armeId}/consommation`,
    payload,
  );
  return data.data;
}

/** Returns the ammunition consumption history for a specific weapon. */
export async function getArmeConsommations(
  armeId: number,
): Promise<ArmeMunitionsConsommation[]> {
  const { data } = await apiClient.get<ApiResponse<ArmeMunitionsConsommation[]>>(
    `/armes/${armeId}/consommations`,
  );
  return data.data;
}

File: android/app/src/main/java/com/gsoft/opus/domain/model/Arme.kt
package com.gsoft.opus.domain.model

/**
 * Weapon type/category (e.g. "Pistolet PA 9mm", "Fusil AK-47").
 * Referenced by [Arme]. Managed via the TypeArme CRUD endpoints.
 */
data class TypeArme(
    val id: Int,
    val nom: String,
    val description: String?,
    /** Shared ammunition stock for all weapons of this type. */
    val munitionsStock: Int,
    val createdAt: String?,
    val updatedAt: String?
)

/**
 * Individual physical weapon identified by its unique matricule.
 * Each arme belongs to a [TypeArme] and tracks its current ammunition
 * stock ([munitionsStock]). The stock is the source of truth for how
 * many rounds are currently available for this weapon; it is decreased
 * atomically by the backend when ammunition is consumed (either via an
 * armement reintegration recording munitionsConsommees, or via the
 * dedicated POST /armes/{id}/consommation endpoint).
 */
data class Arme(
    val id: Int,
    val typeArmeId: Int,
    /** Joined type name from the API. */
    val typeArmeNom: String?,
    val matricule: String,
    /** Legacy per-weapon stock (kept for backward compat, no longer active). */
    val munitionsStock: Int,
    /** Joined from type_arme — the active shared stock for this weapon's type. */
    val typeArmeMunitionsStock: Int,
    val createdAt: String?,
    val updatedAt: String?
)

/**
 * Ammunition consumption history row (auditable log of every ammunition
 * consumption event). Each row records which weapon ([armeId]) was
 * involved, which agent (personnel) consumed the ammunition, how many
 * rounds were consumed, when, and — when the consumption happened during
 * an armement reintegration — which perception ([armementId]) triggered
 * it. This table is the audit trail; the [Arme.munitionsStock] column is
 * the current live stock. The two are kept consistent inside a single
 * database transaction on every consumption.
 */
data class ArmeMunitionsConsommation(
    val id: Int,
    val armeId: Int,
    val agentId: Int?,
    val armementId: Int?,
    val quantite: Int,
    val dateConsommation: String,
    val createdAt: String?,
    /** Joined from arme via the API. */
    val armeMatricule: String?,
    /** Joined from type_arme via the API. */
    val typeArmeNom: String?,
    /** Joined from personnel via the API. */
    val agentIm: String?,
    val agentGrade: String?,
    val agentFirstname: String?,
    val agentLastname: String?
)
